pkg/istio-agent/xds_proxy.go

```diff
@@ -124,69 +129,116 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
+	if p.connected != nil {
+		p.connected.requestsChan <- req
+	}
+}
+
+func (p *XdsProxy) RegisterStream(c *ProxyConnection) {
+	p.connectedMutex.Lock()
+	defer p.connectedMutex.Unlock()
+	if p.connected != nil {
+		close(p.connected.stopChan)
+	}
+	p.connected = c
+}
+
+type ProxyConnection struct {
+	upstreamError   chan error
+	downstreamError chan error
+	requestsChan    chan *discovery.DiscoveryRequest
+	responsesChan   chan *discovery.DiscoveryResponse
+	stopChan        chan struct{}
+	downstream      discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer
+}
+
 // Every time envoy makes a fresh connection to the agent, we reestablish a new connection to the upstream xds
 // This ensures that a new connection between istiod and agent doesn't end up consuming pending messages from envoy
 // as the new connection may not go to the same istiod. Vice versa case also applies.
 func (p *XdsProxy) StreamAggregatedResources(downstream discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
-	proxyLog.Infof("connecting to %s", p.istiodAddress)
+	proxyLog.Infof("Envoy ADS stream established")
+
+	con := &ProxyConnection{
+		upstreamError:   make(chan error),
+		downstreamError: make(chan error),
+		requestsChan:    make(chan *discovery.DiscoveryRequest, 10),
+		responsesChan:   make(chan *discovery.DiscoveryResponse, 10),
+		stopChan:        make(chan struct{}),
+		downstream:      downstream,
+	}
 
-	upstreamError := make(chan error)
-	downstreamError := make(chan error)
-	requestsChan := make(chan *discovery.DiscoveryRequest, 10)
-	responsesChan := make(chan *discovery.DiscoveryResponse, 10)
-	healthEventsChan := make(chan *health.ProbeEvent, 5)
-	// A separate channel for nds requests to not contend with the ones from envoys
-	ndsRequestChan := make(chan *discovery.DiscoveryRequest, 5)
+	p.RegisterStream(con)
 
 	// Handle downstream xds
 	firstNDSSent := false
 	go func() {
-		defer close(requestsChan) // Indicates downstream close.
-		defer close(ndsRequestChan)
 		for {
 			// From Envoy
 			req, err := downstream.Recv()
 			if err != nil {
-				downstreamError <- err
-				close(downstreamError)
+				con.downstreamError <- err
 				return
 			}
 			// forward to istiod
-			requestsChan <- req
+			con.requestsChan <- req
 			if !firstNDSSent && req.TypeUrl == v3.ListenerType {
 				// fire off an initial NDS request
-				ndsRequestChan <- &discovery.DiscoveryRequest{
+				con.requestsChan <- &discovery.DiscoveryRequest{
 					TypeUrl: v3.NameTableType,
 				}
 				firstNDSSent = true
 			}
 		}
 	}()
 
-	// health check stop channel
-	stop := make(chan struct{})
-	defer close(stop)
-	go p.healthChecker.PerformApplicationHealthCheck(healthEventsChan, stop)
-
-	upstreamConn, err := grpc.Dial(p.istiodAddress, p.istiodDialOptions...)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	upstreamConn, err := grpc.DialContext(ctx, p.istiodAddress, p.istiodDialOptions...)
```

Comment:
This still couples the downstream with the upstream

Take a look at https://github.com/istio/istio/pull/27884, we can move it out

Reply:
This is intentional. If you do not couple them, you break the statefulness of upstream and downstream. For example, if upstream disconnects and we reconnect to upstream, then we will not send the full ADS initialization sequence (CDS req, etc), and we will not get updates. Upstream:Downstream must be 1:1

Reply:
It make sense in this initial sequence, but i think lose the initial intention of xds proxy, which is to save connections to istiod

Reply:
The original intent was to have a single connection (before we had one from envoy for XDS and one from agent for CSR) not to to reduce the number of new connections although that is something we can do. I think I was mixing up connection with stream a bit. We MUST be 1:1 on the stream, but maybe we can reuse the connection. We will need to do more research on what happens when istiod closes the connection though if we break it out like https://github.com/istio/istio/pull/27884/files

Reply:
btw, isn't it the same number of connections most of the time? Envoy will never close the stream, so the only time we will reconnect is if the upstream has an error, in which case we need to reconnect anyways. It seems like mostly a refactoring to call `grpc.Dial` once and expect it to continuously reconnect automatically (as done in https://github.com/istio/istio/pull/27884/files) or to directly call `grpc.Dial` as in this PR?

Reply:
It can be a follow-up.